application/F3DStarter.cxx

```diff
@@ -325,112 +326,70 @@ class F3DStarter::F3DInternals
+        std::string formatted = ss.str();
+        if (formatted == fmt)
         {
-          fragments.emplace_back(m, false);
+          f3d::log::warn("invalid date format for \"", var, "\"");
         }
-      };
-
-      const std::regex re(escapedVar + "|" + substVar);
-      std::sregex_token_iterator begin(templateString.begin(), templateString.end(), re, { -1, 0 });
-      std::for_each(begin, std::sregex_token_iterator(), callback);
-
-      return fragments;
+        return formatted;
+      }
+      throw std::out_of_range(var);
     };
 
-    const auto fragments = substituteVariables();
+    f3d::utils::string_template stringTemplate(templateString);
+    stringTemplate.substitute(variableLookup);
 
-    /* check the non-substituted fragments for numbering variables */
     const auto hasNumbering = [&]()
     {
-      for (const auto& [fragment, processed] : fragments)
+      for (const auto& variable : stringTemplate.variables())
       {
-        if (!processed && std::regex_search(fragment, numberingRe))
+        if (std::regex_search(variable, numberingRe))
         {
           return true;
         }
       }
       return false;
     };
 
-    /* just join and return if there's no numbering to be done */
+    /* return if there's no numbering to be done */
     if (!hasNumbering())
     {
-      std::stringstream joined;
-      for (const auto& fragment : fragments)
-      {
-        joined << fragment.first;
-      }
-      return { joined.str() };
+      return { stringTemplate.str() };
     }
 
-    /* apply numbering in the non-substituted fragments and join */
-    const auto applyNumbering = [&](const size_t i)
+    const auto numberingLookup = [&](const size_t i)
     {
-      std::stringstream joined;
-      for (const auto& [fragment, processed] : fragments)
+      return [&numberingRe, i](const std::string& var)
       {
-        if (!processed && std::regex_match(fragment, numberingRe))
+        if (std::regex_match(var, numberingRe))
         {
           std::stringstream formattedNumber;
           try
           {
-            const std::string fmt = std::regex_replace(fragment, numberingRe, "$2");
+            const std::string fmt = std::regex_replace(var, numberingRe, "$2");
             formattedNumber << std::setfill('0') << std::setw(std::stoi(fmt)) << i;
           }
           catch (std::invalid_argument&)
           {
+            if (i == 1) /* avoid spamming the log */
+            {
+              f3d::log::warn("ignoring invalid number format for \"", var, "\"");
+            }
             formattedNumber << std::setw(0) << i;
           }
-          joined << std::regex_replace(fragment, numberingRe, formattedNumber.str());
+          return std::regex_replace(var, numberingRe, formattedNumber.str());
         }
-        else
-        {
-          joined << fragment;
-        }
-      }
-      return joined.str();
+        throw std::out_of_range(var);
```

Comment:
throw f3d specific exception instead

Reply:
Wait, this is in app, do we really need an exception here ?

Reply:
that's the variable lookup function saying "not found" (same way `std::map::at` does) if it doesn't have a value for that key

Reply:
ha right, this is caught in utils.cxx and I see it is documented in utils.h. However, using a std exception means that it is easy to catch an exception we did not intend to catch if at any point the std throw this for some reason (in not yet written code). Using a utils::string_template::exception ensure this will never happen.

Reply:
I thought `std::out_of_range` was more specific than it actually is, oh well custom one it is then